File: algorithmic-toolbox/src/week3/java/BinarySearch.java
package week3.java;
import java.util.Arrays;
import java.util.Scanner;

class BinarySearch {

    public static void main(final String[] args) {
        try (final Scanner in = new Scanner(System.in)) {
            final int[][] as = readArray(in, true), bs = readArray(in, false);
            for (int i = 0; i < bs.length; i++) {
                if (i > 0) {
                    System.out.print(" ");
                }

                final int index = Arrays.binarySearch(as, bs[i], (x, y) -> x[1] - y[1]);
                if (index < 0) {
                    System.out.print("-1");
                } else {
                    System.out.print(as[index][0]);
                }
            }
        }
    }

    private static int[][] readArray(final Scanner in, final boolean sort) {
        final int[][] result = new int[in.nextInt()][];
        for (int i = 0; i < result.length; i++) {
            result[i] = new int[] { i, in.nextInt() };
        }

        if (sort) {
            Arrays.sort(result, (a, b) -> a[1] - b[1]);
        }

        return result;
    }

}

File: algorithmic-toolbox/src/week3/scala/BinarySearch.scala
package week3.scala

import java.util.Scanner

object BinarySearch {
  def main(args: Array[String]) = {
    val in = new Scanner(System.in)
    val a = readList(in).sortBy(_._2)
    println(readList(in).map(n => binarySearch(a, n._2)).mkString(" "))
  }

  def readList(in: Scanner) = (0 until in.nextInt).map((_, in.nextLong))
  def binarySearch(l: Seq[(Int, Long)], k: Long): Int = {
    def binarySearch(lo: Int, hi: Int): Int = {
      val mid = (lo + hi) / 2

      if (lo > hi) -1
      else if (lo == hi) (if (l(lo)._2 == k) l(lo)._1 else -1)
      else if (l(mid)._2 == k) l(mid)._1
      else if ((l(mid)._2 > k) || (mid == lo)) binarySearch(lo, mid) 
      else binarySearch(mid, hi)
    }

    binarySearch(0, l.length)
  }
}